src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/AccessController.cs

```diff
@@ -284,12 +284,12 @@ private async Task<IActionResult> ExchangeClientCredentialsGrantType(OpenIdConne
                     OpenIddictConstants.Destinations.AccessToken,
                     OpenIddictConstants.Destinations.IdentityToken);
 
-                // If the role provider is available, add all the role claims
+                // If the role service is available, add all the role claims
                 // associated with the application roles in the database.
-                var provider = HttpContext.RequestServices.GetService<IRoleProvider>();
-                if (provider != null)
+                var roleService = HttpContext.RequestServices.GetService<IRoleService>();
```

Comment:
Why is this resolved in the loop?

Reply:
This assume the role service could not be registered. But in `ApplicationController` it's injected in the ctor. Is the other controller only accessible if the service is registered?

Reply:
Yes, we should fix `ApplicationController` to ensure `IRoleService` is optional. We can either opt for this approach (service locator) or have 2 constructors (DI is expected to call the one with the most dependencies it can resolve).

Reply:
There is another controller that depends on it.

Reply:
Which one?

Reply:
Thank you, but I have other remarks regarding to this:
1.  You can disable the Roles module, but this will definitely break the site.
2. I think the OrchardCore.OpenId.Management module should depend on the Roles module, because as you said, the `ApplicationController` is only accessible if the service is registered.
3. Yes, we should resolve `IRoleService` outside of the loop.

What do you think? I think I should fill issues about the first two points here.

Reply:
Actually, removing the dependency between the OpenID features and the Users/Roles module was a completely deliberate change, made in https://github.com/OrchardCMS/OrchardCore/pull/1786.

The idea is that we wanted to allow developers to bring their own Users/Roles-like module without being blocked by the fact the OpenID module relied on the built-in ones.

That said, I agree the user experience is not optimal without the dependency relationship and that's why I opened https://github.com/OrchardCMS/OrchardCore/issues/2901 yesterday to track future improvements.